base/task.jl

```diff
@@ -365,6 +365,67 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    foreach(waiting_tasks) do @nospecialize t
+        if !(t isa Task)
+            error("Expected an iterator of `Task` object")
+        end
+    end
+
```

Comment:
Iterating `waiting_tasks` may change the contents (e.g. consuming it) or they may change afterwards, so this isn't sufficient or helpful to check upfront
```suggestion
```

Reply:
Indeed, that would be the case in worst-case scenarios. The intention of this code seems to be to avoid scheduling new `Task` when given invalid `waiting_tasks`. So I'd propose adjusting the main `enumerate` iteration as follows?:
```julia
function _wait_multiple(waiting_tasks; all=false, failfast=false)
    tasks = Task[]
    done_mask = Bool[]
    exception = false
    nremaining::Int = 0

    for (i, t) in enumerate(waiting_tasks)
        t isa Task || error("Expected an iterator of `Task` object")
        push!(tasks, t)
        if istaskdone(t)
            push!(done_mask, true)
            exception |= istaskfailed(t)
        else
            push!(done_mask, false)
            nremaining += 1
        end
    end

    chan = Channel{Tuple{Int,Task}}(Inf)
    for (i, done) in enumerate(done_mask)
        if !done
            t = tasks[i]
            schedule(Task(() -> begin
                _wait(t)
                put!(chan, (i, t))
            end))
        end
    end

    ...
```

Reply:
@aviatesk Thank you for suggesting the code.  This can apply to a consuming task stream and avoid invalid input handling. I'll apply this.

Reply:
I've applied this at 1bd4973.